runtime/src/mm/cpp/ObjectFactory.hpp

```diff
@@ -0,0 +1,295 @@
+/*
+ * Copyright 2010-2020 JetBrains s.r.o. Use of this source code is governed by the Apache 2.0 license
+ * that can be found in the LICENSE file.
+ */
+
+#ifndef RUNTIME_MM_OBJECT_FACTORY_H
+#define RUNTIME_MM_OBJECT_FACTORY_H
+
+#include <algorithm>
+#include <memory>
+#include <mutex>
+
+#include "Alignment.hpp"
+#include "Alloc.h"
+#include "Memory.h"
+#include "Mutex.hpp"
+#include "Utils.hpp"
+
+namespace kotlin {
+namespace mm {
+
+namespace internal {
+
+// A queue that is constructed by collecting subqueues from several `Producer`s.
+// This is essentially a heterogeneous `MultiSourceQueue` on top of a singly linked list that
+// uses `konanAllocMemory` and `konanFreeMemory`
+// TODO: Consider merging with `MultiSourceQueue` somehow.
+class ObjectFactoryStorage : private Pinned {
+public:
+    // This class does not know its size at compile-time.
+    class Node : private Pinned {
+    public:
+        ~Node() = default;
+
+        static void operator delete(void* ptr) noexcept { konanFreeMemory(ptr); }
+
+        // Note: This can only be trivially destructible data, as nobody can invoke its destructor.
+        void* Data() noexcept { return data_; }
+
+    private:
+        friend class ObjectFactoryStorage;
+
+        explicit Node(void* data) noexcept : data_(data) {}
+
+        static std::unique_ptr<Node> Create(size_t dataSize, size_t dataAlignment) noexcept {
+            RuntimeAssert(IsValidAlignment(dataAlignment), "dataAlignment=%zu is not a valid alignment", dataAlignment);
+            RuntimeAssert(IsAligned(dataSize, dataAlignment), "dataSize=%zu must be aligned to dataAlignment=%zu", dataSize, dataAlignment);
+            size_t alignment = std::max(alignof(Node), dataAlignment);
+            size_t size = sizeof(Node) + dataSize;
+            size_t allocSize = AlignUp(size, alignment);
+            void* ptr = konanAllocMemory(allocSize);
+            if (!ptr) {
+                // TODO: Try doing GC first.
+                konan::consoleErrorf("Out of memory trying to allocate %zu. Aborting.\n", allocSize);
+                konan::abort();
+            }
+            void* alignedPtr = AlignUp(ptr, alignment);
```

Comment:
What is this variable required for?

Reply:
I just realized I made a mistake (in `delete`). This code is used to convert a possibly unaligned `ptr` from allocator into a definitely aligned pointer `alignedPtr`. The mistake is that I free `alignedPtr` - not `ptr`.